README.md
## Synopsis

This tool allows you to list contents of all your accounts in a single page (characters, their stats and items, items in vaults). Also it generates a summary of all the items - you probably saw screenshots of these in trading forum ([example](http://i755.photobucket.com/albums/xx195/Ind3sisiv3/Ilovemuledump.png)).

The point of playing the game is to have fun. Muling is not fun. I am trying to increase overall fun ratio by decreasing amount of time spent fussing with mules and storagekeeping.

## Download

All released versions are [here](https://github.com/atomizer/muledump/releases).
Stdman will release as many new versions as is possible [here](https://github.com/thestdman/muledump/releases). 

## Howto

- unpack
- edit **`accounts_sample.js`**
- rename it to **`accounts.js`**
- open **`muledump.html`**

## Howto update your muledump to Stdman's version

- download all files
- copy **accounts.js** file from old muledump
- paste it in the new muledump folder
- voila its really that simple

## Not so obvious features

- click on item to filter accounts that hold it
- click on account name for individual options menu
- ctrl-click account name to temporarily hide it from totals
- read the `accounts.js` file, it has some variables to play with

## Head to the [wiki](https://github.com/atomizer/muledump/wiki) for more information!

And of course, feel free to contribute to it!

## License

Copyright (c) 2013 [atomizer](https://github.com/atomizer)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
**2017-05-31** version 27.7.x.14-1 

- all items since last version added
- updated the update function
- changed vault layout
- fixed invisible vault glitch
- changed the dungeons displayed in additional stats
- added my own touch to it all

**2016-10-16** version 0.6.7

- 27.7.X6 items

**2016-08-31** version 0.6.6

- 27.7.X3 items

**2016-07-25** version 0.6.5

- 27.7.DECA (1469395332) items
- combined stat display (base/max) by [JohnBlackburne](https://github.com/JohnBlackburne)

**2015-11-03** version 0.6.4

- [Kabam.com support](https://github.com/atomizer/muledump/wiki/Kabam-tutorial) (thanks to Zaltais333)
- consumable sorting fixes

**2015-09-22** version 0.6.3

- consumables are now sorted by type (thanks to Pfiffel & Nightfirecat)
- further improvements to the one-click login script (thanks to Nightfirecat)
- [Steam support](https://github.com/atomizer/muledump/wiki/Steam-Tutorial)
- a new item??? (27.7.0)

**2015-04-25** version 0.6.2

- one-click login now writes to both possible locations of the flash cookie
- fixed display of the one-click login links

**2015-02-22** version 0.6.1

- 27.3.1 items
- show all errors at the top of the page (as before)

**2015-01-27** version 0.6.0

- automated migration
- better user experience when mass-reloading
- hopefully will show more info in case of errors

**2015-01-18** version 0.5.1

- fixed cache not updating

**2015-01-17** version 0.5.0

- 27.3.0 items
- proper "migration required" warning
- automated ToS accept and age verification
- fixed PNG export (imgur)
- internal: moved stuff around, might break things
- internal: updated jquery

**2014-10-22** version 0.4.26

- 25.0 items

**2014-09-22** version 0.4.25

- 23.2 items

**2014-07-01** version 0.4.24

- 22.3 items
- one-click login script fix for latest version of AutoIt ([Nightfirecat](https://github.com/Nightfirecat))

**2014-06-09** version 0.4.22

- 21.4 items

really sorry for lack of updates recently :/


**2014-03-24** version 0.4.21

- 19.5 items

**2014-03-01** version 0.4.20

- gift chests option
- artemis skin

**2014-02-21** version 0.4.19

- 19.2 items

**2013-12-19** version 0.4.18

- 19.0 items

**2013-12-13** version 0.4.17

- 18.0 items, small ring positioning tweaks

**2013-09-14** version 0.4.16

- 16.4 items (in advance)
- gift chests ([Nightfirecat](https://github.com/Nightfirecat))
- more tolerant to corrupt char data (zub)
- disabled price lookup since the forums are dead
- email not verified warning

**2013-08-28** version 0.4.15

- 16.0 items
- 16.0 vault chest order ([Nightfirecat](https://github.com/Nightfirecat))
- display characters always in same order ([Nightfirecat](https://github.com/Nightfirecat))
- added `rowlength` variable to accounts.js
- better handling of unknown items (separate counters, item id in hex)
- use DarkDaemon's price guide ([aiedail92](https://github.com/aiedail92))
- updated JQuery and masonry
- more attempts to fix masonry-related rendering bugs

**2013-08-01** version 0.4.14

- 15.0 items
- hp/mp pot counters ([SlugKing](https://github.com/SlugKing), [Nightfirecat](https://github.com/Nightfirecat))
- feed power in tooltips
- copyright and license notes

**2013-03-31** version 0.4.13

- 12.1 items

**2013-02-08** version 0.4.12

- 11.0 items

**2013-01-22** version 0.4.11

- 9.0 items

**2012-12-12** version 0.4.10

- 7.0 items
- changed price guide to MustafaD (pull request by [avoxgames](https://github.com/avoxgames))

**2012-11-16** version 0.4.9

- 5.0 items
- added backpacks
- fixed ninja equips position in totals

**2012-11-2** version 0.4.8

- 4.0 items
- item names and tiers now match the in-game tooltips (except dosed items like elixirs)
- fixed char description sometimes spanning more lines than it should

**2012-10-25** version 0.4.7

- 3.1 items
- made achievement progress and additional stats options independent of each other

**2012-10-11** version 0.4.6

- ninja equipment fame bonuses
- fixed star color calculation

**2012-10-05** version 0.4.5

- ninja, 124.0 items

**2012-09-25** version 0.4.4

- new accounts.js option: "nomasonry" - set to 1 to turn off smart layout
- fixed some problems with per-account menus
- ctrl-click on name toggles the account in the totals

**2012-09-16** version 0.4.3

- authentic character portraits
- fixed long names breaking the box

**2012-09-04** version 0.4.1

- 123.5.0 items

**2012-08-25** version 0.4

- 123.4.4 items
- ability to set options for each account individually - click account name for menu (intended use: keep global options to items only and expand the good accounts as needed)
- added achievement progress calculation (thanks to Pfiffel, even though I didnt use your code)
- vault chests are now in their in-game order (thanks to Hals for assistance)
- improved page load time
- added redirect to Kable's video

**2012-08-05** version 0.3.2

- 123.4.1 items
- reduced probability of duplicate error messages

**2012-07-04** version 0.3.1

- 123.3.2 items
- fixed price guide url
- fixed some visual glitches

**2012-06-13** version 0.3: collaboration edition

- 123.3 stats
- export to TXT, CSV, JSON, PNG
- accuracy and god kill ratio
- fixed "sticky" options bug
- new reload symbol
- optional: one-click login -- by [FizzeBu](http://forums.wildshadow.com/user/24488); more info above
- optional: display prices from Kazansky -- by [aiedail92](https://github.com/aiedail92); more info above

**2012-05-08**

version 0.2.3

- fixed star counter
- better Opera compatibility
- various performance optimizations
- update checking is now on-demand

**2012-05-06** version 0.2.2

- stars
- automatic update checker

**2012-05-02** version 0.2.1

- fixed some visual bugs and edge cases introduced in previous update
- back to old method of issuing requests (slow but more stable)
- you can make options to stay open by clicking
- moved additional stats to the bottom
- updated JQuery

**2012-04-29** version 0.2

- using JQuery Masonry plugin to lay out the accounts

lib/pcstats.js
var pcstatnames = {
	0: 'Shots',
	1: 'ShotsThatDamage',
	2: 'SpecialAbilityUses',
	3: 'TilesUncovered',
	4: 'Teleports',
	5: 'PotionsDrunk',
	6: 'MonsterKills',
	7: 'MonsterAssists',
	8: 'GodKills',
	9: 'GodAssists',
	10: 'CubeKills',
	11: 'OryxKills',
	12: 'QuestsCompleted',
	13: 'PirateCavesCompleted',
	14: 'UDLsCompleted',
	15: 'AbyssCompleted',
	16: 'SnakePitsCompleted',
	17: 'SpiderDensCompleted',
	18: 'SpritesCompleted',
	19: 'LevelUpAssists',
	20: 'MinutesActive',
	21: 'TombsCompleted',
	22: 'TrenchesCompleted',
	23: 'JunglesCompleted',
	24: 'ManorsCompleted',
	25: 'ForestMazesCompleted',
	26: 'EasyLODsCompleted',
	27: '27unknown',
	28: 'CemeteriesCompleted',
	29: 'CaveofTreasureCompleted',
	30: 'MadLabsCompleted',
	31: 'DavysCompleted',
	32: '32unknown',
	33: '33unknown',
	34: 'IceCavesCompleted',
	35: 'DDocksCompleted',
	36: 'CDepthsCompleted',
	37: 'LabyrinthsCompleted',
	38: 'BattleNexusCompleted',
	39: 'ShattersCompleted',
	40: 'BellasCompleted',
	41: 'TheatresCompleted',
	42: 'SewersCompleted',
	43: 'HivesCompleted',
}

var shortdungeonnames = { // sorted by "tier"
	13: 'Pirate',
	23: 'Jungle',
	17: 'Spider',
	16: 'Snake',
	18: 'Sprite',
	24: 'Manor',
	14: 'UDL',
	15: 'Abyss',
	22: 'Trench',
	21: 'Tomb',
}

var bonuses = {
	'Ancestor': function(s, c) {
		return (c.id < 2) ? {mul: 0.1, add: 20} : 0;
	},
	'Legacy Builder': function(s, c, d) {
		// 0.1
	},
	'Pacifist': function(s) {
		return s[1] ? 0 : 0.25;
	},
	'Thirsty': function(s) {
		return s[5] ? 0 : 0.25;
	},
	'Mundane': function(s) {
		return s[2] ? 0 : 0.25;
	},
	'Boots on the Ground': function(s) {
		return s[4] ? 0 : 0.25;
	},
	'Tunnel Rat': function(s) {
		for (var i in shortdungeonnames) if (!s[i]) return 0;
		return 0.1;
	},
	'Enemy of the Gods': function(s) {
		return (s[8] / (s[6] + s[8]) > 0.1) ? 0.1 : 0;
	},
	'Slayer of the Gods': function(s) {
		return (s[8] / (s[6] + s[8]) > 0.5) ? 0.1 : 0;
	},
	'Oryx Slayer': function(s) {
		return s[11] ? 0.1 : 0;
	},
	'Accurate': function(s) {
		return (s[1] / s[0] > 0.25) ? 0.1 : 0;
	},
	'Sharpshooter': function(s) {
		return (s[1] / s[0] > 0.5) ? 0.1 : 0;
	},
	'Sniper': function(s) {
		return (s[1] / s[0] > 0.75) ? 0.1 : 0;
	},
	'Explorer': function(s) {
		return (s[3] > 1e6) ? 0.05 : 0;
	},
	'Cartographer': function(s) {
		return (s[3] > 4e6) ? 0.05 : 0;
	},
	'Team Player': function(s) {
		return (s[19] > 100) ? 0.1 : 0;
	},
	'Leader of Men': function(s) {
		return (s[19] > 1000) ? 0.1 : 0;
	},
	'Doer of Deeds': function(s) {
		return (s[12] > 1000) ? 0.1 : 0;
	},
	'Friend of the Cubes': function(s) {
		return s[10] ? 0 : 0.1;
	},
	'Well Equipped': function(s, c) {
		var eq = c.Equipment.split(',');
		var b = 0;
		for (var i = 0; i < 4; i++) {
			var it = items[+eq[i]] || items[-1];
			b += it[5];
		}
		return b * 0.01;
	},
	'First Born': function(s, c, d, f) {
		return (d.Account.Stats.BestCharFame < f) ? 0.1: 0;
	},
}

var goals = {
	'Tunnel Rat': function(s) {
		var r = [];
		for (var i in shortdungeonnames) {
			if (!s[i]) r.push(shortdungeonnames[i]);
		}
		return [r.join(', '), 'dungeons'];
	},
	'Enemy of the Gods': function(s) {
		var x = s[6] / 9 - s[8];
		if (Math.ceil(x) == x) x += 1;
		return [Math.ceil(x), 'god kills'];
	},
	'Slayer of the Gods': function(s) {
		return [s[6] - s[8] + 1, 'god kills'];
	},
	'Oryx Slayer': function(s) {
		return s[11] ? 0 : [1, 'Oryx kill'];
	},
	'Accurate': function(s) {
		var x = (0.25 * s[0] - s[1]) / 0.75;
		if (Math.ceil(x) == x) x += 1;
		return [Math.ceil(x), 'shots'];
	},
	'Sharpshooter': function(s) {
		var x = (0.5 * s[0] - s[1]) / 0.5;
		if (Math.ceil(x) == x) x += 1;
		return [Math.ceil(x), 'shots'];
	},
	'Sniper': function(s) {
		var x = (0.75 * s[0] - s[1]) / 0.25;
		if (Math.ceil(x) == x) x += 1;
		return [Math.ceil(x), 'shots'];
	},
	'Explorer': function(s) {
		return [1e6 - s[3] + 1, 'tiles'];
	},
	'Cartographer': function(s) {
		return [4e6 - s[3] + 1, 'tiles'];
	},
	'Team Player': function(s) {
		return [100 - s[19] + 1, 'party levelups'];
	},
	'Leader of Men': function(s) {
		return [1000 - s[19] + 1, 'party levelups'];
	},
	'Doer of Deeds': function(s) {
		return [1000 - s[12] + 1, 'quests'];
	},
}


function readstats(pcstats) {
	function readInt32BE(str, idx) {
		var r = 0;
		for (var i = 0; i < 4; i++) {
			var t = str.charCodeAt(idx + 3 - i);
			r += t << (8 * i);
		}
		return r;
	}

	pcstats = pcstats || ''
	var b = atob(pcstats.replace(/-/g, '+').replace(/_/g, '/'));
	var r = [];
	for (var i = 0; i < b.length; i += 5) {
		var f = b.charCodeAt(i);
		var val = readInt32BE(b, i + 1);
		r[f] = val;
	}
	for (var i in pcstatnames) if (!r[i]) r[i] = 0;
	return r;
}

function printstats(c, d, dogoals, dostats) {
	var st = readstats(c.PCStats);
	var $c = $('<table class="pcstats" />');
	var fame = +c.CurrentFame;
	
	function tline(name, val, cl) {
		$('<tr>')
			.append($('<td>').text(name))
			.append($('<td>').addClass(cl || 'pcstat').text(val))
			.appendTo($c);
	}
	function gline(t, b) {
		$('<tr>')
			.append($('<td colspan=2>')
				.addClass('goal')
				.append($('<span>').text(t))
				.append($('<span class="bonus">').text(b))).appendTo($c);
	}
	
	if (dogoals) {
		var p = '';
		for (var a in goals) {
			var x = goals[a](st);
			if (!x || x[0] <= 0) continue;
			var s = x[0] + ' for ';
			if (p != x[1]) {
				p = x[1]
				s = '\u2022 ' + x[1] + ': ' + s
			}
			gline(s, a);
		}
	}
	if (!dostats) return $c;
	
	for (var i in st) {
		if (!st[i]) continue;
		var sname = pcstatnames[i] || '#' + i;
		tline(sname, st[i]);
	}
	if (st[20] > 59) {
		var v = st[20], r = []
		var divs = { 'd': 24 * 60, 'h': 60, 'm': 1 }
		for (var s in divs) {
			if (r.length > 2) break;
			var t = Math.floor(v / divs[s]);
			if (t) r.push(t + s);
			v %= divs[s];
		}
		tline('Active', r.join(' '), 'info');
	}
	if (st[0] && st[1]) {
		tline('Accuracy', Math.round(10000 * st[1] / st[0]) / 100 + '%', 'info');
	}
	if (st[8]) {
		tline('God kill ratio', Math.round(10000 * st[8] / (st[6] + st[8])) / 100 + '%', 'info');
	}
	
	if (!fame) return $c;
	for (var k in bonuses) {
		var b = bonuses[k](st, c, d, fame);
		if (!b) continue;
		var incr = 0;
		if (typeof b == 'object') {
			incr += b.add;
			b = b.mul;
		}
		incr += Math.floor(fame * b);
		fame += incr;
		tline(k, '+' + incr, 'bonus');
	}
	tline('Total Fame', fame, 'bonus');
	
	return $c;
}

